### ISSUE TITLE
twisted.names.dns.Name loops indefinitely when encountering a compression loop

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5064]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5064|
|Type|defect|
|Created|2011-04-23 21:23:12Z|
|Branch|https://github.com/twisted/twisted/tree/dns-comp-loop-5064|

Twisted Names is vulnerable to a fairly trivial, (at least somewhat) well known (eg CVE-2007-3409) attack.  The attack involves constructing a DNS name which includes a label which is a pointer to itself or an earlier label in the name ending with the pointer.  Twisted Names will follow this loop forever (or until it runs out of memory, at least).

<details><summary>Searchable metadata</summary>

```
trac-id__5064 5064
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_dns_comp_loop_5064 branches/dns-comp-loop-5064
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__names names
keywords__security security
time__1303593792000000 1303593792000000
changetime__1350578620000000 1350578620000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

See bzr+ssh://svn.twistedmatrix.com/~exarkun/dns-comp-loop-5064/

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

Reviewing.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

Sigh.  Attempting to pull this branch into my shared repository results in this error -  https://bugs.launchpad.net/bzr/+bug/769973 - so, "this bug affects you".

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun<br>@glyph set status to `new`|
|-|-|

`twisted.python.compat.set` is not necessary, as we no longer support python 2.3.  [http://docs.python.org/release/2.4/whatsnew/node2.html The `set` builtin was introduced in 2.4.]

Delete the import for that and then please land.  (dpush to a regular branch, then use SVN to do the merge; the bug above confirms my reticence to let any more bzr metadata into the repository.)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [31716]) Branching to 'dns-comp-loop-5064'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [31719]) Branching to 'dns-comp-loop-5064'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [31721]) Merge dns-comp-loop-5064

Author: exarkun
Reviewer: glyph, zooko
Fixes: [#5064](https://github.com/twisted/twisted/issues/5064)

Add compressed name loop detection to the Twisted Names decoder so that when such a
loop is encountered, processing stops and an exception is raised.  Previously such a
loop would lead to an infinite loop in the decoder.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

Couldn't dpush, <https://bugs.launchpad.net/ubuntu/+source/bzr-svn/+bug/769992>, bzr diff/patched instead.
